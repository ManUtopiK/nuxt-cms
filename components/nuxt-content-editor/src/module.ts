import { defineNuxtModule, addPlugin, addComponent, addPrerenderRoutes, addServerHandler, createResolver, installModule } from '@nuxt/kit'

// Module options TypeScript interface definition
export interface ModuleOptions { }

export default defineNuxtModule<ModuleOptions>({
  meta: {
    name: 'nuxt-content-editor',
    configKey: 'nuxtContentEditor'
  },
  // Default configuration options of the Nuxt module
  defaults: {},
  async setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url)

    // Do not add the extension since the `.ts` will be transpiled to `.mjs` after `npm run prepack`
    addPlugin(resolve('./runtime/plugin'))

    addComponent({ name: 'ContentEditor', filePath: resolve('./runtime/components/Editor.vue') })
    // addImports({ name: 'useMonaco', as: 'useMonaco', from: resolve('composables') })

    addServerHandler({
      method: "get",
      route: "/__nuxt_content_editor.json",
      handler: resolve("./runtime/server/routes/nuxt_content_editor")
    })
    addPrerenderRoutes("/__nuxt_content_editor.json")
    await installModule("nuxt-component-meta", {
      globalsOnly: true
    })
  }
})
