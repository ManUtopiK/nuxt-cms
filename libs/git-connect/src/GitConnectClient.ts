import GitAuthClient from './lib/GitAuthClient'
import { GitConnectClientOptions } from './types'

export default class GitConnectClient {
  auth: GitAuthClient
  api: string
  options: GitConnectClientOptions

  constructor(options: GitConnectClientOptions) {
    // TODO Detect github and gitlab, or provider and dispatch GitAuth...
    this.auth = new GitAuthClient({ ...options })
    this.api = this.auth.config.api
    Object.assign(this, options)
  }

  async query({ query, variables }: { query: any, variables?: object }) {
    const results = await fetch(`${this.api}/api/graphql`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: await this.auth.getToken(),
      },
      body: JSON.stringify({ query, ...(variables && { variables }) }),
    })
    return await results.json()
  }
}
