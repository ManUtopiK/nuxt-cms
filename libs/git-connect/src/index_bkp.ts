import { ref } from 'vue'
import {
  NetlifyAuthenticator,
  ImplicitAuthenticator,
  PkceAuthenticator,
} from 'netlify-cms-lib-auth'

const clientSideAuthenticators = {
  pkce: ({ base_url, auth_endpoint, app_id, auth_token_endpoint }) =>
    new PkceAuthenticator({
      base_url,
      auth_endpoint,
      app_id,
      auth_token_endpoint,
    }),

  implicit: ({ base_url, auth_endpoint, app_id, clearHash }) =>
    new ImplicitAuthenticator({ base_url, auth_endpoint, app_id, clearHash }),

  netlify: ({ base_url, auth_endpoint, site_id }) =>
    new NetlifyAuthenticator({
      base_url,
      auth_endpoint,
      site_id,
    }),
}

const auth = ref()
let api_url

type Args = {
  auth_type?: string
  base_url?: string
  auth_endpoint?: string
  app_id?: string
}

export const useGitX = ({
  auth_type = 'pkce',
  base_url = 'https://gitlab.com',
  auth_endpoint = 'oauth/authorize',
  app_id = '',
}: Args = {}) => {
  const Authenticator = clientSideAuthenticators[auth_type]({
    base_url,
    auth_endpoint,
    app_id,
    auth_token_endpoint: 'oauth/token',
  })

  const authenticate = () => {
    Authenticator.authenticate({ provider: 'gitlab', scope: 'api' })
  }

  const completeAuth = () => {
    api_url = base_url
    return new Promise((resolve, reject) => {
      Authenticator.completeAuth((err, data) => {
        if (err) {
          reject(err)
        }
        auth.value = data
        resolve(auth)
      })
    })
  }

  const query = async (query) => {
    const results = await fetch(`${api_url}/api/graphql`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `${auth.value.token_type} ${auth.value.access_token}`,
      },
      body: JSON.stringify({
        query,
      }),
    })
    return (await results.json()).data
  }

  return { authenticate, completeAuth, auth, query, Authenticator }
}
