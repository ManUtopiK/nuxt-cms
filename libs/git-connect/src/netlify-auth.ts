const NETLIFY_API = 'https://api.netlify.com'
const AUTH_ENDPOINT = 'auth'

function trim(str, ch) {
  let start = 0
  let end = str.length

  while (start < end && str[start] === ch) {
    ++start
  }

  while (end > start && str[end - 1] === ch) {
    --end
  }

  return start > 0 || end < str.length ? str.substring(start, end) : str
}

class NetlifyError {
  err: {
    message: String
  }

  constructor(err) {
    this.err = err
  }

  toString() {
    return this.err && this.err.message
  }
}

const PROVIDERS = {
  github: {
    width: 960,
    height: 600,
  },
  gitlab: {
    width: 960,
    height: 600,
  },
  bitbucket: {
    width: 960,
    height: 500,
  },
  email: {
    width: 500,
    height: 400,
  },
}

class Authenticator {
  siteId: String
  baseUrl: String
  authEndpoint: String
  authWindow: {
    postMessage: Function
    close: Function
    focus: Function
  }

  constructor(config = {}) {
    this.siteId = config.siteId || null
    this.baseUrl = trim(config.baseUrl, '/') || NETLIFY_API
    this.authEndpoint = trim(config.authEndpoint, '/') || AUTH_ENDPOINT
  }

  handshakeCallback(options, cb) {
    const fn = (e) => {
      if (
        e.data === 'authorizing:' + options.provider &&
        e.origin === this.baseUrl
      ) {
        window.removeEventListener('message', fn, false)
        window.addEventListener(
          'message',
          this.authorizeCallback(options, cb),
          false
        )
        return this.authWindow.postMessage(e.data, e.origin)
      }
    }
    return fn
  }

  authorizeCallback(options, cb) {
    const fn = (e) => {
      if (e.origin !== this.baseUrl) {
        return
      }

      if (
        e.data.indexOf('authorization:' + options.provider + ':success:') === 0
      ) {
        const data = JSON.parse(
          e.data.match(
            new RegExp('^authorization:' + options.provider + ':success:(.+)$')
          )[1]
        )
        window.removeEventListener('message', fn, false)
        this.authWindow.close()
        cb(null, data)
      }
      if (
        e.data.indexOf('authorization:' + options.provider + ':error:') === 0
      ) {
        const err = JSON.parse(
          e.data.match(
            new RegExp('^authorization:' + options.provider + ':error:(.+)$')
          )[1]
        )
        window.removeEventListener('message', fn, false)
        this.authWindow.close()
        cb(new NetlifyError(err))
      }
    }
    return fn
  }

  getSiteID() {
    if (this.siteId) {
      return this.siteId
    }
    const host = document.location.host.split(':')[0]
    return host === 'localhost' ? 'cms.netlify.com' : host
  }

  authenticate(options, cb) {
    const { provider } = options
    const siteID = this.getSiteID()

    if (!provider) {
      return cb(
        new NetlifyError({
          message:
            'You must specify a provider when calling netlify.authenticate',
        })
      )
    }
    if (!siteID) {
      return cb(
        new NetlifyError({
          message:
            "You must set a siteId with netlify.configure({siteId: 'your-site-id'}) to make authentication work from localhost",
        })
      )
    }

    const conf = PROVIDERS[provider] || PROVIDERS.github
    const left = screen.width / 2 - conf.width / 2
    const top = screen.height / 2 - conf.height / 2
    window.addEventListener(
      'message',
      this.handshakeCallback(options, cb),
      false
    )
    let url = `${this.baseUrl}/${this.authEndpoint}?provider=${options.provider}&siteId=${siteID}`
    if (options.scope) {
      url += '&scope=' + options.scope
    }
    if (options.login === true) {
      url += '&login=true'
    }
    if (options.beta_invite) {
      url += '&beta_invite=' + options.beta_invite
    }
    if (options.invite_code) {
      url += '&invite_code=' + options.invite_code
    }
    this.authWindow = window.open(
      url,
      'Netlify Authorization',
      `width=${conf.width}, height=${conf.height}, top=${top}, left=${left}`
    )
    this.authWindow.focus()
  }

  refresh(options, cb) {
    const { provider, refresh_token } = options
    const siteID = this.getSiteID()
    const onError = cb || Promise.reject.bind(Promise)

    if (!provider || !refresh_token) {
      return onError(
        new NetlifyError({
          message:
            'You must specify a provider and refresh token when calling netlify.refresh',
        })
      )
    }
    if (!siteID) {
      return onError(
        new NetlifyError({
          message:
            "You must set a siteId with netlify.configure({siteId: 'your-site-id'}) to make token refresh work from localhost",
        })
      )
    }
    const url = `${this.baseUrl}/${this.authEndpoint}/refresh?provider=${provider}&siteId=${siteID}&refresh_token=${refresh_token}`
    const refreshPromise = fetch(url, { method: 'POST', body: '' }).then(
      (res) => res.json()
    )

    // Return a promise if a callback wasn't provided
    if (!cb) {
      return refreshPromise
    }

    // Otherwise, use the provided callback.
    refreshPromise.then((data) => cb(null, data)).catch(cb)
  }
}

export default Authenticator
