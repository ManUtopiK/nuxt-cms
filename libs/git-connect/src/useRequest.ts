import { reactive, computed } from 'vue'

const activeRequests = reactive([])

export default function usePageRequests() {
  const isLoading = computed(() => !!activeRequests.length)
  const makeRequest = async (url) => {
    // push the url to the activeRequests
    const index = activeRequests.length
    activeRequests[index] = url
    const response = await fetch(url).catch((error) => alert(error)) // if failed remove the url from the activeRequests and alert error
    const data = await response.json()
    // remove the url from activeRequests
    activeRequests.splice(index, 1)
    return data
  }

  return { isLoading, makeRequest }
}
