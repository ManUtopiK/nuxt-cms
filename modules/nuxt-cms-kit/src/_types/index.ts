import './hooks'

export * from './custom-tabs'
export * from './client-api'
export * from './server-ctx'
export * from './module-options'
