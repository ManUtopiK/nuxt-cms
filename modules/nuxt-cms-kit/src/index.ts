import { useNuxt } from '@nuxt/kit'
import type { ModuleCustomTab } from './types'

/**
 * Hooks to extend a custom tab in devtools.
 *
 * Provide a function to pass a factory that can be updated dynamically.
 */
export function addCustomTab (tab: ModuleCustomTab | (() => ModuleCustomTab | Promise<ModuleCustomTab>), nuxt = useNuxt()) {
  nuxt.hook('cms:customTabs', async (tabs: ModuleCustomTab[]) => {
    if (typeof tab === 'function') { tab = await tab() }
    tabs.push(tab)
  })
}

/**
 * Retrigger update for custom tabs, `cms:customTabs` will be called again.
 */
export function refreshCustomTabs (nuxt = useNuxt()) {
  return nuxt.callHook('cms:customTabs:refresh')
}
