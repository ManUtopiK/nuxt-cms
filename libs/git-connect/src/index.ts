import GitConnectClient from './GitConnectClient'
import { GitConnectClientOptions } from './types'

export { default as GitConnectClient } from './GitConnectClient'
export type { GitConnectClientOptions } from './types'

export const createClient = (options: GitConnectClientOptions): GitConnectClient => {
  // Skip if it's running in node.js
  // if (typeof window === 'undefined') { return }

  return new GitConnectClient(options)
}
