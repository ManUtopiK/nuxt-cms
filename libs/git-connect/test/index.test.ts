import { describe, expect, it } from 'vitest'
import { createClient } from '../src'

describe('Its createClient', () => {
  const gitlab = createClient({
    api: 'https://gitlab.com',
    appId: 'xxx'
  })

  it('returns gitlab GitAuthClient', () => {
    expect(gitlab).toMatchSnapshot()
  })

  const github = createClient({
    api: 'https://github.com',
    appId: 'xxx'
  })

  it('returns github GitAuthClient', () => {
    expect(github).toMatchSnapshot()
  })
})
