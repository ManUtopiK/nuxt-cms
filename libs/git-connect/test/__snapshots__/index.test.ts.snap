// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Its createClient > returns github GitAuthClient 1`] = `
GitConnectClient {
  "api": "https://github.com",
  "appId": "xxx",
  "auth": GitAuthClient {
    "Authenticator": PkceAuthenticator {
      "appId": "xxx",
      "authTokenUrl": "https://github.com/oauth/token",
      "authUrl": "https://github.com/oauth/authorize",
      "redirectPath": "/git-connect/redirect",
    },
    "access_token": undefined,
    "config": {
      "api": "https://github.com",
      "appId": "xxx",
      "authEndpoint": "oauth/authorize",
      "authTokenEndpoint": "oauth/token",
      "authType": "pkce",
      "provider": "gitlab",
      "redirectPath": "/git-connect/redirect",
    },
    "session": undefined,
    "token_type": undefined,
  },
  "options": undefined,
}
`;

exports[`Its createClient > returns gitlab GitAuthClient 1`] = `
GitConnectClient {
  "api": "https://gitlab.com",
  "appId": "xxx",
  "auth": GitAuthClient {
    "Authenticator": PkceAuthenticator {
      "appId": "xxx",
      "authTokenUrl": "https://gitlab.com/oauth/token",
      "authUrl": "https://gitlab.com/oauth/authorize",
      "redirectPath": "/git-connect/redirect",
    },
    "access_token": undefined,
    "config": {
      "api": "https://gitlab.com",
      "appId": "xxx",
      "authEndpoint": "oauth/authorize",
      "authTokenEndpoint": "oauth/token",
      "authType": "pkce",
      "provider": "gitlab",
      "redirectPath": "/git-connect/redirect",
    },
    "session": undefined,
    "token_type": undefined,
  },
  "options": undefined,
}
`;
